File: internal/lcu/cmdline.go
package lcu

// 命令行检测通道（兜底，覆盖 WeGame 等注册表不可靠场景）：
// 枚举 LeagueClientUx 进程，从命令行提取 --app-port / --remoting-auth-token / --rso-platform-id。

import (
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"sync"

	"github.com/shirou/gopsutil/v4/process"
)

// 预编译正则（命令行高频解析，避免重复编译）。
// token/platform 值允许紧跟引号（"--remoting-auth-token=xxx" 形态），捕获组跳过引号取值。
var (
	reAppPort     = regexp.MustCompile(`--app-port=(\d+)`)
	reAuthToken   = regexp.MustCompile(`--remoting-auth-token="?([^\s"']+)`)
	rePlatformID  = regexp.MustCompile(`--rso_platform_id="?([^\s"']+)`)
	leagueProcSet = map[string]bool{"leagueclientux.exe": true, "leagueclientux": true}
)

// ProcessScan 进程扫描结果
type ProcessScan struct {
	PID        int32
	Port       uint16
	Token      string
	PlatformID string
	ExeDir     string
}

// ParseCmdline 从 LeagueClientUx 命令行提取连接参数（纯函数，可测）。
// ok=true 表示 port 与 token 均提取成功；platformID 可为空。
func ParseCmdline(cmd string) (port uint16, token string, platformID string, ok bool) {
	m := reAppPort.FindStringSubmatch(cmd)
	if m == nil {
		return
	}
	p64, err := strconv.ParseUint(m[1], 10, 16)
	if err != nil || p64 == 0 {
		return
	}
	t := reAuthToken.FindStringSubmatch(cmd)
	if t == nil || t[1] == "" {
		return
	}
	port = uint16(p64)
	token = t[1]
	if s := rePlatformID.FindStringSubmatch(cmd); s != nil {
		platformID = s[1]
	}
	ok = true
	return
}

// SanitizeCmdline 日志脱敏：token 打码后输出（凭据不得进日志）
func SanitizeCmdline(cmd string) string {
	return reAuthToken.ReplaceAllString(cmd, "--remoting-auth-token=***")
}

// ScanLeagueClientUx 枚举 LeagueClientUx 进程，返回首个凭据齐备的结果。
// 同名但无凭据的进程（僵尸/无权限）跳过继续扫描，不提前退出。
func ScanLeagueClientUx() (*ProcessScan, bool) {
	pids, err := process.Pids()
	if err != nil {
		return nil, false
	}
	for _, pid := range pids {
		p, err := process.NewProcess(pid)
		if err != nil {
			continue
		}
		name, err := p.Name()
		if err != nil || !leagueProcSet[strings.ToLower(name)] {
			continue
		}
		cmd, _ := p.Cmdline()
		if cmd == "" {
			continue
		}
		port, token, platformID, ok := ParseCmdline(cmd)
		if !ok {
			continue
		}
		scan := &ProcessScan{PID: pid, Port: port, Token: token, PlatformID: platformID}
		if exe, err := p.Exe(); err == nil {
			scan.ExeDir = filepath.Dir(exe)
		}
		return scan, true
	}
	return nil, false
}

// PidAlive 进程存活检查（lockfile 残留校验）
func PidAlive(pid int32) bool {
	ok, err := process.PidExists(pid)
	return err == nil && ok
}

// 大区标识缓存：lockfile 通道不含 --rso_platform_id，需命令行补充；
// 按 PID 缓存避免每轮轮询全量扫描进程。
var (
	platCacheMu  sync.Mutex
	platCachePID int32
	platCacheID  string
)

// platformIDForPID 返回指定客户端进程的大区标识（带缓存；找不到返回空串）
func platformIDForPID(pid int32) string {
	platCacheMu.Lock()
	if platCachePID == pid && platCacheID != "" {
		id := platCacheID
		platCacheMu.Unlock()
		return id
	}
	platCacheMu.Unlock()

	id := scanPlatformID()
	if id != "" {
		platCacheMu.Lock()
		platCachePID, platCacheID = pid, id
		platCacheMu.Unlock()
	}
	return id
}

// scanPlatformID 从 LeagueClientUx 命令行提取 --rso_platform_id（登录大区标识）
func scanPlatformID() string {
	pids, err := process.Pids()
	if err != nil {
		return ""
	}
	for _, pid := range pids {
		p, err := process.NewProcess(pid)
		if err != nil {
			continue
		}
		name, err := p.Name()
		if err != nil || !leagueProcSet[strings.ToLower(name)] {
			continue
		}
		cmd, _ := p.Cmdline()
		if m := rePlatformID.FindStringSubmatch(cmd); m != nil {
			return m[1]
		}
	}
	return ""
}

File: src-tauri/src/lcu/cmdline.rs
//! 进程扫描：定位 LeagueClient 凭据（对齐 Go internal/lcu/cmdline.go）。
//! 通道 A：lockfile（注册表 Location / 用户配置目录 + PID 存活）。
//! 通道 B：LeagueClientUx 命令行 --app-port / --remoting-auth-token（国服/WeGame 主路径）。

use std::sync::Mutex;

use sysinfo::{ProcessRefreshKind, ProcessesToUpdate, System, UpdateKind};

use super::lockfile::{lockfile_candidates, parse_cmdline_overrides, parse_lockfile, pid_alive};
use super::types::Credentials;

/// 大区标识缓存（按 PID，避免每轮全量扫 Ux）。
static PLAT_CACHE: Mutex<Option<(i32, String)>> = Mutex::new(None);

/// 扫描系统，找到 LCU 凭据。
/// 优先级：lockfile 通道 > LeagueClientUx 命令行通道。
pub fn detect_credentials(client_path: &str) -> Option<Credentials> {
    if let Some(c) = detect_from_lockfile(client_path) {
        log::info!("[lcu] lockfile channel hit pid={} port={}", c.pid, c.port);
        return Some(c);
    }
    if let Some(c) = detect_from_ux_cmdline() {
        log::info!(
            "[lcu] cmdline channel hit pid={} port={} platform={}",
            c.pid,
            c.port,
            c.platform_id
        );
        return Some(c);
    }
    None
}

/// 通道 A：候选 lockfile 路径解析 + PID 存活校验（对齐 Go readLockfile）。
/// 国服 lockfile 常为空文件——parse 失败即跳过，不视为命中。
fn detect_from_lockfile(client_path: &str) -> Option<Credentials> {
    for path in lockfile_candidates(client_path) {
        let Ok(content) = std::fs::read_to_string(&path) else {
            continue;
        };
        let Some(mut c) = parse_lockfile(&content) else {
            continue;
        };
        if !pid_alive(c.pid) {
            // 残留 lockfile：跳过
            continue;
        }
        c.platform_id = platform_id_for_pid(c.pid);
        return Some(c);
    }
    None
}

/// 通道 B：枚举 LeagueClientUx，命令行提取凭据（对齐 Go ScanLeagueClientUx）。
/// sysinfo 的 cmd 在国服常为空 → 优先 NT API `ProcessCommandLineInformation`。
fn detect_from_ux_cmdline() -> Option<Credentials> {
    for (pid, cmd) in enumerate_ux_cmdlines() {
        if let Some((port, token, platform)) = parse_cmdline_overrides(&cmd) {
            let platform_id = platform.unwrap_or_else(|| platform_id_for_pid(pid));
            return Some(Credentials {
                pid,
                port,
                token,
                platform_id,
            });
        }
    }
    None
}

/// 枚举 LeagueClientUx 的 (pid, cmdline)；cmd 走 sysinfo → NT API 回退。
fn enumerate_ux_cmdlines() -> Vec<(i32, String)> {
    let mut out = Vec::new();
    let mut sys = System::new();
    sys.refresh_processes_specifics(
        ProcessesToUpdate::All,
        true,
        ProcessRefreshKind::nothing().with_exe(UpdateKind::OnlyIfNotSet),
    );

    for (pid, proc_) in sys.processes() {
        let name = proc_.name().to_string_lossy().to_lowercase();
        if name != "leagueclientux.exe" && name != "leagueclientux" {
            continue;
        }
        let pid_i = pid.as_u32() as i32;
        let mut cmd = cmd_to_string(proc_.cmd());
        if cmd.is_empty() {
            #[cfg(windows)]
            {
                cmd = super::win_cmdline::read_command_line(pid.as_u32()).unwrap_or_default();
            }
        }
        if !cmd.is_empty() {
            out.push((pid_i, cmd));
        }
    }
    out
}

/// 从 Ux 命令行提取 `--rso_platform_id`（带 PID 缓存，对齐 Go platformIDForPID）。
fn platform_id_for_pid(pid: i32) -> String {
    if let Ok(cache) = PLAT_CACHE.lock() {
        if let Some((p, id)) = cache.as_ref() {
            if *p == pid && !id.is_empty() {
                return id.clone();
            }
        }
    }

    let id = scan_platform_id().unwrap_or_default();
    if !id.is_empty() {
        if let Ok(mut cache) = PLAT_CACHE.lock() {
            *cache = Some((pid, id.clone()));
        }
    }
    id
}

fn scan_platform_id() -> Option<String> {
    for (_, cmd) in enumerate_ux_cmdlines() {
        if let Some((_, _, Some(p))) = parse_cmdline_overrides(&cmd) {
            return Some(p);
        }
        if let Some(idx) = cmd.find("--rso_platform_id=") {
            let rest = &cmd[idx + "--rso_platform_id=".len()..];
            let end = rest.find(char::is_whitespace).unwrap_or(rest.len());
            let v = rest[..end].trim_matches('"');
            if !v.is_empty() {
                return Some(v.to_string());
            }
        }
    }
    None
}

fn cmd_to_string(cmd: &[std::ffi::OsString]) -> String {
    cmd.iter()
        .map(|s| s.to_string_lossy().into_owned())
        .collect::<Vec<_>>()
        .join(" ")
}

/// 单测辅助：清空平台缓存。
#[cfg(test)]
pub fn clear_platform_cache() {
    if let Ok(mut c) = PLAT_CACHE.lock() {
        *c = None;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// 真机冒烟：Ux 进程在跑时必须命中；未跑则跳过。
    #[test]
    fn detect_live_when_client_running() {
        let ux_alive = {
            let mut sys = System::new();
            sys.refresh_processes_specifics(
                ProcessesToUpdate::All,
                true,
                ProcessRefreshKind::nothing(),
            );
            sys.processes().values().any(|p| {
                let n = p.name().to_string_lossy().to_lowercase();
                n == "leagueclientux.exe" || n == "leagueclientux"
            })
        };
        if !ux_alive {
            eprintln!("skip: LeagueClientUx not running");
            return;
        }
        let c = detect_credentials("").expect("Ux running must detect credentials");
        assert!(c.port > 0);
        assert!(!c.token.is_empty());
        assert!(c.pid > 0);
        eprintln!(
            "HIT pid={} port={} platform={}",
            c.pid, c.port, c.platform_id
        );
    }

    #[cfg(windows)]
    #[test]
    fn nt_reads_own_cmdline() {
        let s =
            super::super::win_cmdline::read_command_line(std::process::id()).expect("nt cmdline");
        assert!(!s.is_empty());
    }
}
